design/pool-configuration.md

```diff
@@ -0,0 +1,386 @@
+* [PR: MetalLB speakers working on a subset of nodes in layer2](https://github.com/metallb/metallb/pull/613)
+* [Issue: Annotate nodes with address pool/weighted address pools](https://github.com/metallb/metallb/issues/941)
+
+The following issues also fall in the same "speaker announcement logic" category:
+
+Service with external backends
+
+* [Issue: LoadBalancer service is not accessible using external IP, when endpoints (with a pod inside the cluster) is created manually](https://github.com/metallb/metallb/issues/717)
+* [Issue: Announce route for nodeless endpoints](https://github.com/metallb/metallb/issues/646)
+
+### IP Announcement Targets
+
+Currently, an IP assigned to a Service is advertised to all the possible destinations from all the nodes.
+The only exception is the BGP peer [node selector](https://metallb.universe.tf/configuration/#limiting-peers-to-certain-nodes).
+
+Here, we suggest to extend the configuration logic with a protocol dependent way to select the target(s) of the announcement.
+
+#### BGP Announcement logic
+
+Different IPs belonging to different pools may need to be announced to different subset of BGP peers, associated to different networks.
+
+This is described in:
+
+* [Issue: BGP peers per address-pool](https://github.com/metallb/metallb/issues/521)
+* [PR: Limit announcement of address pools to specific BGP peers](https://github.com/metallb/metallb/pull/596)
+
+Here we want to choose what BGP peers a node should advertise the IP of the pool to:
+
+[<img src="./images/pool_peer_selector.svg" width="400"/>](./images/pool_peer_selector.svg)
+
+#### L2 Announcement logic
+
+The L2 speaker currently responds to any request coming from connected interfaces (and sends gratuitous arps on all the connected interfaces when it needs to).
+
+We could think about selectively choose what interfaces / clients send the replies / broadcasts to, but given the locality nature of the L2 mode, plus the fact that
+no issues were raised for this scenario, this option won't be covered in this proposal.
+
+#### No announcement
+
+This third (currently not implemented) mode raised in [this discussion](https://github.com/metallb/metallb/pull/832#issuecomment-888897400) where
+there are scenarios in which only the ip pooling side of metallb is used, but the routes to those ips are set statically.
+
+### Mixing nodes / target selection
+
+A more complete scenario would mix the two configuration degrees described above, advertising the same pool to different peers depending
+on the node the ip is being advertised from.
+
+[<img src="./images/pool_peer_node_selector.svg" width="400"/>](./images/pool_peer_node_selector.svg)
+
+No issues were raised but this combination is a more flexible variant of the [peer node selector](https://metallb.universe.tf/configuration/#limiting-peers-to-certain-nodes).
+
+With this combination, the user may choose a different set of targets depending on the node.
```

Comment:
So node allocation and peer selection are kind of orthogonal. The exceptions are due to a node being unable to reach a peer. So where would we check for those exceptions? The webhook that you mention below?

Reply:
I think that is clearer in the examples below

There are two types of node selection here:

* The pool node selector drives which nodes are allowed to advertise the ip
* The node selector in the bgp peer of a given pool, tells that the ip must be advertised to that particular peer from a given subset of nodes

A webhook would be able to spot configurations with no intersections between the two selectors resulting in a pool that won't be advertised anywhere.